pkg/services/ngalert/remote/remote_primary_forked_alertmanager.go

```diff
@@ -76,10 +76,20 @@ func (fam *RemotePrimaryForkedAlertmanager) TestTemplate(ctx context.Context, c
 	return fam.remote.TestTemplate(ctx, c)
 }
 
-func (fam *RemotePrimaryForkedAlertmanager) CleanUp() {}
+func (fam *RemotePrimaryForkedAlertmanager) CleanUp() {
+	// No cleanup to do in the remote Alertmanager.
+	fam.internal.CleanUp()
+}
 
-func (fam *RemotePrimaryForkedAlertmanager) StopAndWait() {}
+func (fam *RemotePrimaryForkedAlertmanager) StopAndWait() {
+	fam.internal.StopAndWait()
```

Comment:
In that case why we would stop the internal sender? It shouldn't be running in remote primary mode, should it?

Reply:
That's shutting down the entire internal Alertmanager, not the sender.
- `fam.internal.StopAndWait()` will shut down the internal Alertmanager's components (inhibitor, dispatcher, channels it needs in order to run).
- `fam.remote.StopAndWait()` will shut down the sender we have running for forwarding alerts to the remote Alertmanager.

We need the internal Alertmanager running even though we're not sending alerts there. We use the internal Alertmanager's `StopAndWait()` method to shut it down as a whole.

Reply:
Both the dispatcher and the inhibitor are started in the internal Alertmanager's `ApplyConfig()` method, the code is in our [alerting repository](https://github.com/grafana/alerting/blob/bf12694896a843b7a77d7e7ef656013334214ef9/notify/grafana_alertmanager.go#L427-L437). We need to shut these down when exiting Grafana, and we can't start the internal Alertmanager without running these components.